### DISCUSSION TITLE
CMake preset for sysbuild

### DISCUSSION THREAD

**Discussion by vChavezB** [zephyrproject-rtos/zephyr#75466] [Q&A] [answered]
Does anyone know if CMake presets can be used with sysbuild ?

`west build MY_COOL_PROJECT -b MY_BOARD --sysbuild -- --preset MY_PRESET`

The problem is that sysbuild looks for CMakePresets.json in zephyr/share/sysbuild  instead of `MY_COOL_PROJECT/CMakePresets.json`.

`CMake Error: Could not read presets from C:/SOME_DIR/workspace/zephyr/share/sysbuild `

**Comment by soghoyan** [comment]
Actually it's possible with cmake. But would require to use a wrapper cmake project, which adds zephyr/share/sysbuild as subdirectory, and some small changes in zephyr/share/sysbuild/cmake/modules/sysbuild_extensions.cmake to make sysbuild working as non-root project.
If someone still needs this, I can describe in details. I used sysbuild with presets for integration with VSCode CMake Tools.

**Reply by vChavezB** [reply]
I would be interested in knowing about this approach 👍

**Reply by soghoyan** [reply]
It's turned out there is a way to build zephyr project with sysbuild and cmake presets without any changes in baseline. Below  is the configuration which works for me.
Directory structure:
```
YourProject
   |
   |-CmakeLists.txt
   |-CMakePresets.json // for regular build
   |-sysbuild-wrapper
      |
      |-CmakeLists.txt
      |-CMakePresets.json
      |-image_configurations
         |-...
```
Inside `sysbuild-wrapper`:
`CmakeLists.txt`:
```
cmake_minimum_required(VERSION 3.20)

project(sysbuild-wrapper)

add_subdirectory(<path-to-zephyr>/share/sysbuild sysbuild-wrapper)
```
`CMakePresets.json`:
```
{
    "version": 8,
    "configurePresets": [
        {
            "name": "sysbuild",
            "displayName": "Zephyr Sysbuild",
            "description": "Zephyr Sysbuild Configure",
            "binaryDir": "${sourceDir}/../../../out/build/${sourceDirName}",
            "generator": "Ninja",
            "cacheVariables": {
                "APP_DIR": "${sourceDir}/../",
                "BOARD": "<YourBoardID>"
            },
            "environment": {
            }
        }
    ],
    "buildPresets": [
        {
            "name": "sysbuild",
            "displayName": "Zephyr Sysbuild",
            "description": "Zephyr Sysbuild Build",
            "configurePreset": "sysbuild",
            "inheritConfigureEnvironment": true,
            "environment": {
            },
            "targets": [
                "all"
            ]
        }
    ]
}
```
Copy all `.cmake` files from `zephyr/share/sysbuild/image_configurations` to `YourProject/sysbuild-wrapper/image_configurations`.
Now when you call cmake from `YourProject/sysbuild-wrapper` and specify preset `sysbuild` it should properly build an app with bootloader.

The issue comes from the following commands in`zephyr/share/sysbuild/cmake/modules/sysbuild_extensions.cmake`:
```
set(image_default "${CMAKE_SOURCE_DIR}/image_configurations/ALL_image_default.cmake")

if(DEFINED ZBUILD_APP_TYPE)
   list(APPEND image_default "${CMAKE_SOURCE_DIR}/image_configurations/${ZBUILD_APP_TYPE}_image_default.cmake")
endif()
```
This is the only place where root cmake project path is referenced, so replacing it with `${ZEPHYR_BASE}/share/sysbuild/image_configurations/` would do the trick without copy of `image_configurations` to the project.

From the VSCode integration point of view I found this solution useless: although it allows to do sysbuild using CMake  Tools, but features like Intellisense or variable expansion according to active project/build are not available :(

I hope this would be helpful for your case...

**Reply by tejlmand** [reply, marked as answer]
Official CMake preset support for sysbuild was introduced with https://github.com/zephyrproject-rtos/zephyr/pull/96609
and is being improved here: https://github.com/zephyrproject-rtos/zephyr/pull/120249

Documentation is found here: https://docs.zephyrproject.org/latest/build/sysbuild/index.html#sysbuild-and-cmake-presets
